MuxExoPlayer/src/main/java/com/mux/stats/sdk/muxstats/AdsImaSDKListener.java

```diff
@@ -23,53 +26,71 @@ public AdsImaSDKListener(MuxBaseExoPlayer listener) {
+        }
+    }
+
+    private void setupAdViewData(PlaybackEvent event, Ad ad) {
+        ViewData viewData = new ViewData();
+        if (exoPlayerListener.getCurrentPosition() == 0) {
+            if (ad != null) {
+                viewData.setViewPrerollAdId(ad.getAdId());
+                viewData.setViewPrerollCreativeId(ad.getCreativeId());
+            }
+        }
+        event.setViewData(viewData);
     }
 
     @Override
     public void onAdEvent(AdEvent adEvent) {
         if (exoPlayerListener != null) {
+            PlaybackEvent event;
+            Ad ad = adEvent.getAd();
             switch (adEvent.getType()) {
-                case AD_BREAK_STARTED:
-                    exoPlayerListener.dispatch(new AdBreakStartEvent(null));
+                case CONTENT_PAUSE_REQUESTED:
+                    event = new AdBreakStartEvent(null);
+                    setupAdViewData(event, ad);
+                    exoPlayerListener.dispatch(event);
+                    event = new AdRequestEvent(null);
                     break;
-                case AD_BREAK_ENDED:
-                    exoPlayerListener.dispatch(new AdBreakEndEvent(null));
+                case CONTENT_RESUME_REQUESTED:
+                    event = new AdBreakEndEvent(null);
+                    break;
+                case LOADED:
+                    event = new AdResponseEvent(null);
+                    setupAdViewData(event, ad);
+                    exoPlayerListener.dispatch(event);
+                    event = new AdPlayEvent(null);
                     break;
                 case STARTED:
-                    exoPlayerListener.dispatch(new AdPlayEvent(null));
-                    exoPlayerListener.dispatch(new AdPlayingEvent(null));
+                    event = new AdPlayingEvent(null);
                     break;
                 case FIRST_QUARTILE:
-                    exoPlayerListener.dispatch(new AdFirstQuartileEvent(null));
+                    event = new AdFirstQuartileEvent(null);
                     break;
                 case MIDPOINT:
-                    exoPlayerListener.dispatch(new AdMidpointEvent(null));
+                    event = new AdMidpointEvent(null);
                     break;
                 case THIRD_QUARTILE:
-                    exoPlayerListener.dispatch(new AdThirdQuartileEvent(null));
+                    event = new AdThirdQuartileEvent(null);
                     break;
                 case COMPLETED:
-                    exoPlayerListener.dispatch(new AdEndedEvent(null));
+                    event = new AdEndedEvent(null);
                     break;
                 case PAUSED:
-                    exoPlayerListener.dispatch(new AdPauseEvent(null));
+                    event = new AdPauseEvent(null);
                     break;
                 case RESUMED:
-                    exoPlayerListener.dispatch(new AdPlayingEvent(null));
+                    event = new AdPlayingEvent(null);
```

Comment:
We need an AdPlay event before this AdPlaying event, so that internally we know the state.